### PR TITLE
Feature/smhe 1882 emi detection eventlog reading

### PR DESCRIPTION
**Pull request by harrymiddelburgetall**
Added reading of the events from the POWER_QUALITY_THD_EVENT_LOG
- THD_OVER_LIMIT_L1
- THD_OVER_LIMIT_L2
- THD_OVER_LIMIT_L3
- THD_NORMAL_L1
- THD_NORMAL_L2
- THD_NORMAL_L3
- THD_LONG_OVER_LIMIT_L1
- THD_LONG_OVER_LIMIT_L2
- THD_LONG_OVER_LIMIT_L3
Events have a timestamp. code and magnitude.
Object definition is implemented in the ObjectConfig.
SOURCE_OBJECTS property is added to CosemObject where applicable. This enables DATA objects to reference REGISTER objects for their scaler_unit. This is also implemented for POWER_QUALITY_EXTENDED_EVENT_LOG

### CODE REVIEW

File: osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/objectconfig/DlmsProfileValidator.java [hunk 1]
```diff
@@ -247,6 +249,23 @@ private static void allCaptureObjectsShouldExist(
...
+        dlmsProfile.getObjects().stream()
+            .filter(object -> object.getClassId() == InterfaceClass.DATA.id())
+            .map(ObjectConfigService::getSourceObjectDefinitions)
+            .flatMap(Collection::stream)
+            .map(ObjectConfigService::getCosemObjectTypeFromCaptureObjectDefinition)
+            .map(tag -> objectShouldExist(dlmsProfile, tag))
+            .filter(error -> !error.isEmpty())
+            .collect(Collectors.joining(", "));
+
+    if (!validationError.isEmpty()) {
```
Comment (natasja-n) [inline osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/objectconfig/DlmsProfileValidator.java:264]: Deze if wordt niet gedekt in de unit tests.

File: osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/services/ObjectConfigService.java [hunk 2]
```diff
@@ -313,6 +313,22 @@ public static List<String> getCaptureObjectDefinitions(final CosemObject profile
     return List.of(captureObjectDefinitions);
   }
 
+  public static List<String> getSourceObjectDefinitions(final CosemObject profile) {
```
Comment (natasja-n) [inline osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/services/ObjectConfigService.java:316]: I expect specific unit tests for this new method.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java [hunk 3]
```diff
@@ -123,36 +140,99 @@ private Integer extractEventCounter(
...
+      final String scalerUnitDuration =
+          this.getScalerUnit(
+              eventCode, protocol, DlmsObjectType.POWER_QUALITY_EXTENDED_EVENT_DURATION);
+
+      this.checkIsNumber(eventData.get(3), DURATION);
+      eventDetails.add(
+          new EventDetailDto(
+              DURATION, this.readValueWithScalerAndUnit(eventData.get(3), scalerUnitDuration)));
+    } catch (final ObjectConfigException e) {
+      throw new ProtocolAdapterException("Error in object config", e);
```
Comment (natasja-n) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java]: In general, we don't pay much attention in checking exceptions in the unit tests. I think we should allways conver them.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java [hunk 4]
```diff
@@ -123,36 +140,99 @@ private Integer extractEventCounter(
...
+      throws ObjectConfigException {
+    final CosemObject eventMagnitude =
+        this.objectConfigService.getCosemObject(
+            protocol.getName(), protocol.getVersion(), dlmsObjectType);
+    final Map<String, String> eventMapping =
+        (Map<String, String>) eventMagnitude.getProperties().get(ObjectProperty.SOURCE_OBJECTS);
+    final String sourceObjectName = eventMapping.get(String.valueOf(eventCode));
+
+    final CosemObject cosemSourceObject =
+        this.objectConfigService.getCosemObject(
```
Comment (natasja-n) [inline osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java]: return immediately

File: osgp/protocol-adapter-dlms/osgp-protocol-simulator-dlms/simulator/dlms-device-simulator/src/main/java/org/opensmartgridplatform/simulator/protocol/dlms/cosem/PowerQualityThdEventLog.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            new CaptureObject(1, "0.0.96.11.22.255", (byte) 2, 0), new UInteger16DataProcessor()));
+
+    return definitions;
+  }
+
+  /** Initializes buffer with some data. */
+  private void initBufferData() {
+    this.bufferData = new CircularFifoQueue<>(PROFILE_ENTRIES);
+
+    // Add all events.
```
Comment (natasja-n) [inline osgp/protocol-adapter-dlms/osgp-protocol-simulator-dlms/simulator/dlms-device-simulator/src/main/java/org/opensmartgridplatform/simulator/protocol/dlms/cosem/PowerQualityThdEventLog.java]: comments really necessary?

Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-protocol-simulator-dlms/simulator/dlms-device-simulator/src/main/java/org/opensmartgridplatform/simulator/protocol/dlms/cosem/PowerQualityThdEventLog.java]: Removed commentline. Made method documentation more descriptive. Renamed parameters of method addEvents more descriptive.

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java (hunk 4 above)
Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java]: Yes, I will do that

File: osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java (hunk 3 above)
Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-protocol-adapter-dlms/src/main/java/org/opensmartgridplatform/adapter/protocol/dlms/application/mapping/DataObjectToEventListConverter.java]: Added exception checking with more specific messgages on deeper level methods. And added Unit test checking these exceptions accordingly.
The last exception catch in method getEventRelatedCosemSourceObject is actually checked in the cosem ObjectConfig and is not tested in unit test separately.

File: osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/services/ObjectConfigService.java (hunk 2 above)
Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/services/ObjectConfigService.java:316]: This method is tested in the  DlmsProfileValidatorTest

File: osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/objectconfig/DlmsProfileValidator.java (hunk 1 above)
Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/objectconfig/DlmsProfileValidator.java:264]: This test checks the validity the the DLMS profile and should test positive compile time. The definition does not change runtime. If validator errors exist they should be fixed designtime

File: osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/services/ObjectConfigService.java (hunk 2 above)
Comment (natasja-n) [reply osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/services/ObjectConfigService.java:316]: Unit tests for this public method in the ObjectConfigServiceTest would be more clearer and more direct.

**Review by natasja-n — DISMISSED** [review]

File: osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/services/ObjectConfigService.java (hunk 2 above)
Comment (harrymiddelburgetall) [reply osgp/protocol-adapter-dlms/osgp-dlms/src/main/java/org/opensmartgridplatform/dlms/services/ObjectConfigService.java:316]: Okay, you are right, I will add test for this method in ObjectConfigServiceTest

**Review by natasja-n — APPROVED** [review]

**Pull request merged**